btree/btree.c

```diff
@@ -3213,12 +3213,12 @@ static void ff_cap_reg_define(struct nd *tgt, int idx, uint8_t flags)
 
 	if ((flags & CR_HDR_MASK) == CR_HDR_FULL) {
 		tmp.ncr_hdr_beg = h;
-		tmp.ncr_hdr_end = h->ff_opaque == NULL ?
-				  tmp.ncr_hdr_beg + sizeof(*h) :
-				  (void *)(&h->ff_opaque) - 1;
+		tmp.ncr_hdr_end = (void *)(&h->ff_opaque);
 	} else {
+		/* Capture footer along with ff_var. */
 		tmp.ncr_hdr_beg = &h->ff_var;
-		tmp.ncr_hdr_end = tmp.ncr_hdr_beg + sizeof(struct ff_hdr_var);
+		tmp.ncr_hdr_end = tmp.ncr_hdr_beg + sizeof(h->ff_var) +
+				  sizeof(h->ff_foot);
```

Comment:
This is a bit dangerous since you are assuming that there will be no variables between ff_var and ff_foot. In future this might not be true. A safer approach would be
```suggestion
		tmp.ncr_hdr_end = tmp.ncr_hdr_beg + offsetof(h->ff_var) +
				  offsetof(h->ff_foot) + sizeof(h->ff_foot);
```

I assume there is a macro/compiler keyword offsetof. Please confirm.

Reply:
This is by design. We plan to keep the frequently changing members together in ff_head.
If any new member is to be added which needs to be frequently captured, it should be added in struct ff_var.
Added a comment in the structure definition.

Reply:
I mean there could be a case where a non-frequently changing member could be listed between ff_var and ff_foot. There is no hard-coded rule which dictates that the ff_var should always be right before the ff_foot member. If that is supposed to be the case then please add a compile-time assert to make sure that ff_var is always right before ff_foot. You can add a compile-time assert such as the one below right after the location where struct ff_head is declared in the code.
M0_BASSERT(offsetof(struct ff_head, ff_var) + sizeof(((struct ff_head *)0)->ff_var) == offsetof(struct ff_head, ff_foot));